Add estimate_uncertainty_retro wrapper function

seabbs opened the pull request:
**This is entirely from an agent but I have manually reviewed and it seems reasonable. There is some potential argument both of these functions could/should pass `delay_pmf`.  I can't remember the previous discussion about that. The arg is that this functionality is about uncertainty and so should be optionally independent from estimating the delay. The counter arg is that it adds complexity. I think I slightly favour making it an option especially as I expect to reduce arg complexity by refatoring the uncertainty model parameters. If we do want that I am happy to fold it in to this PR.**

## Summary

Adds `estimate_uncertainty_retro()` wrapper function that chains the retrospective nowcasting workflow for uncertainty estimation. Improves API consistency and reduces code duplication.

## Changes

### New wrapper function: `estimate_uncertainty_retro()`
Automates the retrospective nowcasting workflow:
1. `truncate_triangles()` - Creates retrospective snapshots
2. `construct_triangles()` - Generates retrospective reporting triangles  
3. `fill_triangles()` - Generates point nowcasts
4. `estimate_uncertainty()` - Estimates uncertainty parameters

### Key features
- Uses `...` to pass arguments (`ref_time_aggregator`, `delay_aggregator`, `uncertainty_model`) to `estimate_uncertainty()`
- Explicit `structure` parameter (needed by `construct_triangles()`)
- Consistent parameter naming with `estimate_and_apply_uncertainty()`:
  - `n_history_delay` - number of observations for delay estimation
  - `n_retrospective_nowcasts` - number of retrospective snapshots
- Both parameters required (no defaults) to encourage explicit specification

### Refactoring
- `estimate_and_apply_uncertainty()` now uses `estimate_uncertainty_retro()` internally
- Removed duplicate validation between wrappers
- Added `structure` parameter to both wrappers

### Documentation improvements
- Changed "retrospective validation" → "retrospective nowcasts" for accuracy
- Added signposting to low-level functions in higher-level wrapper
- Comprehensive test coverage (23 tests) with descriptive titles

Closes #266

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added `estimate_uncertainty_retro()` function for retrospective uncertainty parameter estimation in nowcasting workflows.
  * Added `structure` parameter to `estimate_and_apply_uncertainty()` for configurable column grouping (default: 1).

* **Documentation**
  * Updated documentation to reflect the new retrospective uncertainty estimation workflow and clarify function relationships.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seabbs commented:
Updated to make `n_retro` a required parameter (no default) to align with the design pattern in `estimate_and_apply_uncertainty`. This ensures users explicitly specify the number of retrospective nowcasts to avoid surprising compute/memory usage.

Also created follow-up issue #296 to refactor `estimate_and_apply_uncertainty` to use this new wrapper function.

kaitejohnson commented on R/estimate_and_apply_uncertainty.R:
```diff
@@ -58,33 +76,12 @@ estimate_and_apply_uncertainty <- function(
-  .validate_inputs_uncertainty(
-    n_min_delay = min_ref_times_delay,
-    n_ref_times = n_ref_times,
+  uncertainty_params <- estimate_uncertainty_retro(
```
@seabbs isn't this #296 ?

seabbs commented on R/estimate_uncertainty_retro.R:
```diff
@@ -0,0 +1,131 @@
+#'   max_delay = 3
+#' )
+#' uncertainty_params_custom
+estimate_uncertainty_retro <- function(
```
This was drawn from the nomenclature vignette. It is hard to find something that is short enough.

The key I think is that this is retrospective on a single triangle. Whereas with the estimate_uncertainty you have lots of choices of what goes in.

I don't think that we should change the underlying function as it is accurate for the functionality and more general hence having a more general name.

seabbs replied:
If this resolves as wanting to change the underlying function name I think that might need to be a new issue

seabbs commented:
@coderabbitai review

seabbs commented on R/estimate_and_apply_uncertainty.R:
```diff
@@ -1,5 +1,22 @@
+#'   \item [sample_nowcasts()] - Applies uncertainty to generate draws
+#' }
+#'
+#' For uncertainty estimation without sampling, use
```
See: [79bcdc2](https://github.com/epinowcast/baselinenowcast/pull/295/commits/79bcdc20f264b5f9d48ba6d16c660b76e5af2cf6)

seabbs commented:
## Added delay_pmf parameter support

We have added support for passing a custom delay PMF (probability mass function) to both `estimate_uncertainty_retro()` and `estimate_and_apply_uncertainty()`.

### Changes Made

1. **`estimate_uncertainty_retro()`**: Added `delay_pmf` parameter that is passed through to `fill_triangles()`
2. **`estimate_and_apply_uncertainty()`**: Added `delay_pmf` parameter that is passed through to `estimate_uncertainty_retro()`
3. **Tests**: Added test case `estimate_uncertainty_retro works with custom delay_pmf` that verifies:
   - Custom delay PMF produces valid results
   - Results differ from default behaviour (where delay is estimated from data)

### API Usage

Users can now provide a custom delay distribution:

```r
# Use custom delay PMF
custom_delay <- c(0.4, 0.3, 0.2, 0.1)

# In estimate_uncertainty_retro
uncertainty_params <- estimate_uncertainty_retro(
  triangle,
  n_history_delay = 5,
  n_retrospective_nowcasts = 2,
  delay_pmf = custom_delay
)

# In estimate_and_apply_uncertainty
nowcast_draws <- estimate_and_apply_uncertainty(
  pt_nowcast_matrix,
  triangle,
  n_history_delay = 4,
  n_retrospective_nowcasts = 2,
  delay_pmf = custom_delay
)
```

All tests pass (27/27).

kaitejohnson commented on tests/testthat/test-estimate_uncertainty_retro.R:
```diff
@@ -0,0 +1,405 @@
+    n_retrospective_nowcasts = 2,
+    delay_pmf = custom_delay_pmf
+  )
+
```
Could add a check for a too short or too long delay_pmf?

seabbs commented:
One option is to use custom expectations everywhere a lot more which might do the trick

kaitejohnson reviewed (approved):
As discussed seems to be a separate issue to test that you get the correct error message from the high level functions

kaitejohnson commented:
> If happy the regexp of low level issues is something for a wider package review I think this is good to go.

By low level issues you mean low level functions?
